### PR TITLE
Roaming lane [WIP]

### LINKED ISSUE #1411: Add roaming to lane position
Should be easy to figure out. If the ratio of the mode to the rest is at some threshold (close to 1:1) the hero is considered roaming.

Thoughts?

### PR DESCRIPTION
**Pull request by masad-frost**
resolves #1411

### CODE REVIEW

**Comment by masad-frost** [comment]
I'm not sure if this is the right approach. I'm thinking maybe we should leave the lane intact and return a roaming bool i.e `getLaneFromPosData`  -> `{ lane, lane_role, is_roaming }`

**Comment by howardchung** [comment]
yeah, I agree that would be less risky, since we retain all of the current data and can choose whether to use the new property or not.  It'll require a schema update though.

**Comment by masad-frost** [comment]
Yeah I figured. And on the UI, we can do Mid (Roaming) or something, which gives more information

**Comment by masad-frost** [comment]
Then again, it will add complexity to filter and counts if we decide to include roaming in there

**Comment by masad-frost** [comment]
I added player is roaming to player caches for future filtering.

`ALTER TABLE player_caches ADD is_roaming text;`

File: sql/create_tables.cql [hunk 1]
```diff
@@ -139,6 +139,7 @@ CREATE TABLE player_caches (
   neutral_kills text,
   lane text,
   lane_role text,
+  is_roaming text,
```
Comment (howardchung) [inline sql/create_tables.cql]: would this also be useful in the SQL database for pro matches?

File: util/utility.js [hunk 2]
```diff
@@ -273,9 +273,9 @@ function mergeObjects(merge, val) {
 }
 
 /**
- * Finds the mode of the input array
+ * Finds the mode and its occurances of the input array
  **/
-function mode(array) {
+function modeWithCount(array) {
```
Comment (howardchung) [inline util/utility.js]: could we preserve the existing function API?  I like having a "mode" function that just returns the mode.  We could make it take an options object to return the additional data.

File: util/utility.js [hunk 3]
```diff
@@ -726,6 +728,7 @@ function getLaneFromPosData(lanePos, isRadiant) {
   return {
     lane,
     lane_role: laneRoles[lane],
+    is_roaming: (count / lanes.length) < 0.45,
```
Comment (howardchung) [inline util/utility.js]: maybe add a comment to explain what this is doing

File: sql/create_tables.cql (hunk 1 above)
Comment (masad-frost) [inline sql/create_tables.cql]: I thought only raw data (`lane_pos`) is stored on that one.

Comment (howardchung) [inline sql/create_tables.cql]: Yeah, but we could add a field.  I suppose that it should probably be part of a different effort/PR to add the lane_role/lane/is_roaming properties for easier querying/filtering.

File: util/utility.js [hunk 4]
```diff
@@ -292,6 +292,10 @@ function modeWithCount(array) {
   return { mode: maxEl, count: maxCount };
 }
 
+function mode(array) {
+  return modeWithCount(array).mode;
```
Comment (masad-frost) [inline util/utility.js]: Is this ok? I didn't do the option because I'm usually against the same function returning 2 different types.

**Comment by masad-frost** [comment]
@howardchung make sure you take care of the migration before deploying

**Comment by howardchung** [comment]
I already did

**Pull request merged**